### PR TITLE
refactor(batch,agg): refactor batch selective agg

### LINKED ISSUE #4759: bug: FILTER seems to be ignored in combination with STRING_AGG
**Describe the bug**
FILTER seems to be ignored in combination with STRING_AGG

**To Reproduce**
```
dev=> CREATE TABLE t(g INT, x VARCHAR, y VARCHAR);
CREATE_TABLE
dev=> INSERT INTO t VALUES (1, 'a', '+'), (2, 'b', '+'), (1, 'c', '-'), (1, 'd', '-');
INSERT 0 4
dev=> SELECT y, STRING_AGG(x, ',' ORDER BY x) FILTER (WHERE g < 2)
    FROM t
    GROUP BY y
    ORDER BY y;
 y | ?column? 
---+----------
 + | a,b // <-- 'b' is related to g == 2 and thus should be filtered out.
 - | c,d
(2 rows)

```

**Expected behavior**
```
dev=> SELECT y, STRING_AGG(x, ',' ORDER BY x) FILTER (WHERE g < 2)
    FROM t
    GROUP BY y
    ORDER BY y;
 y | ?column? 
---+----------
 + | a
 - | c,d
```

### PR DESCRIPTION
**Pull request by richardchien**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

This PR reimplement filter as a standalone `Aggregator` that wraps a child aggregator, so that we don't have to do the filtering in every aggregator. This also fixes #4759.

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Refer to a related PR or issue link (optional)

#4759

### CODE REVIEW

File: src/expr/src/vector_op/agg/filter.rs [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+use crate::Result;
+
+#[derive(Clone)]
+pub struct Filter {
+    filter: ExpressionRef,
+    child: BoxedAggState,
+}
+
+impl Filter {
+    pub fn new(filter: ExpressionRef, child: BoxedAggState) -> Self {
```
Comment (xxchan) [inline src/expr/src/vector_op/agg/filter.rs:31]: Call it `child` is slightly misleading, which looks like the filter is applied to the output of the agg 😄. 

I'm thinking about something like `agg.with_filter(filter)`, but maybe simply adding some docs to `Filter` is enough.

File: src/expr/src/vector_op/agg/filter.rs [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+        self.child.return_type()
+    }
+
+    fn update_single(&mut self, input: &DataChunk, row_id: usize) -> Result<()> {
+        let (row_ref, vis) = input.row_at(row_id)?;
+        assert!(vis);
+        if self
+            .filter
+            .eval_row(&row_ref.to_owned_row())?
+            .expect("filter must have non-null result")
```
Comment (xxchan) [inline src/expr/src/vector_op/agg/filter.rs:48]: filter can produce null, e.g., `=null` 🥶

```sql
dev=> create table t(x int);
CREATE_TABLE
dev=> insert into t values (1);
INSERT 0 1
dev=> select sum(x) filter (where x=null) from t group by x;
server closed the connection unexpectedly
	This probably means the server terminated abnormally
	before or while processing the request.
The connection to the server was lost. Attempting reset: Failed.
```

File: src/expr/src/vector_op/agg/filter.rs [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+            self.child.update_multi(input, start_row_id, end_row_id)
+        } else {
+            for (_, row_id) in (start_row_id..end_row_id)
+                .enumerate()
+                .filter(|(i, _)| bitmap.is_set(*i).unwrap())
+            {
+                self.child.update_single(input, row_id)?;
+            }
+            Ok(())
+        }
```
Comment (xxchan) [inline src/expr/src/vector_op/agg/filter.rs:91]: Would the following method work?
```rust
        let filtered = input.with_visibility(bitmap);
        self.child.update_multi(&filtered, start_row_id, end_row_id)
```

Comment (richardchien) [reply src/expr/src/vector_op/agg/filter.rs:91]: that would require a clone of the whole chunk

Comment (xxchan) [reply src/expr/src/vector_op/agg/filter.rs:91]: `DataChunk` owns `Column`, which owns `ArrayRef`, so can be cheaply cloned.

Comment (richardchien) [reply src/expr/src/vector_op/agg/filter.rs:91]: Seems not that easy. Batch agg states are wrote with the assumption that the data chunk is compacted, which means all rows are visible. While chunk compaction, which does array compaction underlying, is not as cheap as copying a reference.

To remove the assumption on input data chunk, more code refactoring will be required. I think it's a better idea to leave this as future optimization.

**Review by xxchan — APPROVED** [review]

**Pull request merged**